Marking an entity as clean after the afterSave event.

lorenzo opened the pull request:
By letting the entity be dirty during the event, listeners
can inspect the changes done to the entity during the save process.
I think this is a very important information to have and does not
negatively affect anything else.

lorenzo committed:
Marking an enty as clean after the afterSave event.

By letting the entity be dirty during the event, listeners
can inspect the changes done to the entity during the save process.
I think this is a very important information to have and does not
negatively affect anything else.

ADmad commented:
@lorenzo Can the save operation be done on a non-dirty entity?

lorenzo commented:
@ADmad no, save() stops the process if the entity is clean

ADmad commented:
@lorenzo That's what I thought. So one can always assume it to be dirty in an afterSave() callback. How does this change actually help?

I am not against it, just trying to understand.

dakota commented:
Without this change, you wouldn't be able to determine which fields changed/were dirty in `afterSave()`?

lorenzo commented:
@dakota correct, which makes building an audit behavior a sad journey

ADmad commented:
@dakota Exactly. For some reason I was thinking of "dirty" as state of entity as whole instead of individual fields. It's clear to me now :)

ADmad commented:
:+1:

markstory commented:
Seems like a good idea. Making it easy to build audit trails is totally something we should be doing.

lorenzo committed:
Avoding wastful UPDATE statements for HasMany associations

This prevents the ORM from issueing an update statement to the
database if no actual changes could be found

lorenzo commented:
Added another commit that prevents having dirty HasMany entities when they are not actually dirty.

lorenzo committed:
Fixed indentation

markstory commented:
Nice work on finding the bonus UPDATE statements. :clap:
